build_tools/hack/env_check/device.py

```diff
@@ -0,0 +1,631 @@
+                    ("ullAvailPhys", ctypes.c_ulonglong),
+                    ("ullTotalPageFile", ctypes.c_ulonglong),
+                    ("ullAvailPageFile", ctypes.c_ulonglong),
+                    ("ullTotalVirtual", ctypes.c_ulonglong),
+                    ("ullAvailVirtual", ctypes.c_ulonglong),
+                    ("sullAvailExtendedVirtual", ctypes.c_ulonglong),
+                ]
+
+            mem_status = memSTAT()
+            mem_status.dwLength = ctypes.sizeof(memSTAT())
+
+            ctypes.windll.kernel32.GlobalMemoryStatusEx(ctypes.byref(mem_status))
+
+            MEM_PHYS_TOTAL, MEM_PHYS_AVAIL, MEM_VITURAL_AVAIL = (
+                float(mem_status.ullTotalPhys / (1024**3)),
+                float(mem_status.ullAvailPhys / (1024**3)),
+                float(mem_status.ullAvailPageFile / (1024**3)),
+            )
+
+            return (MEM_PHYS_TOTAL, MEM_PHYS_AVAIL, MEM_VITURAL_AVAIL)
+        elif self.is_linux:
+            import re
+
+            with open("/proc/meminfo", "r") as f:
+                _f = f.read().splitlines()
+                for line in _f:
+                    mem_tol = re.search(r"^MemTotal:\s+(\d+)\s+kB$", line)
+                    mem_avl = re.search(r"^MemAvailable:\s+(\d+)\s+kB$", line)
+                    mem_swp = re.search(r"^SwapTotal:\s+(\d+)\s+kB$", line)
+
+                    if mem_tol:
+                        MEM_PHYS_TOTAL = float(mem_tol.group(1)) / (1024**2)
+                    elif mem_avl:
+                        MEM_PHYS_AVAIL = float(mem_avl.group(1)) / (1024**2)
+                    elif mem_swp:
+                        MEM_SWAP_AVAIL = float(mem_swp.group(1)) / (1024**2)
+
+            return (MEM_PHYS_TOTAL, MEM_PHYS_AVAIL, MEM_SWAP_AVAIL)
+        else:
+            return None
+        ...
+
+    def device_disk_status(self):
+        """
+        Return a tuple with Disk Total/Usage messages.
+        `(DISK_DEVICE, DISK_MOUNT_POINT, DISK_TOTAL_SPACE, DISK_USAGE_SPACE, DISK_AVAIL_SPACE, DISK_USAGE_RATIO)`
+        - `DISK_DEVICE`: Returns `str`. The device "contains this repo" name and its mounting point.
+         - Windows: Returns a Drive Letter. eg `F:/` or `F:`
+         - Linux: Returns disk's mounted device name and its mounting point. eg `/dev/sdd at: /`
+        - `DISK_REPO_POINT`: Returns `str`. TheRock current repo abs path.
+        - `DISK_TOTAL_SPACE`: Returns `float`. Current repo stored disk's total space.
+        - `DISK_USAGE_SPACE`: Returns `float`. Current repo stored disk's used space.
+        - `DISK_AVAIL_SPACE`: Returns `float`. Current repo stored disk's avail space.
+        - `DISK_USAGE_RATIO`: Returns `float`. Current repo stored disk's current usage percentage.
+        """
+
+        import os, subprocess
+        from shutil import disk_usage
+        from pathlib import Path
+
+        if self.is_windows:
+            repo_path = TheRock.repo()
+            repo_disk = os.path.splitdrive(repo_path)[0]
```

Comment:
nit: could use pathlib `.drive` instead of `os.path`: https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.drive